arbitrator/prover/src/merkle.rs

```diff
@@ -393,20 +371,39 @@ impl Merkle {
             );
         }
         let mut layers = self.layers.lock();
+        let start = layers.data[0].len();
         let mut layer_size = new_len;
         for (layer_i, layer) in layers.data.iter_mut().enumerate() {
             layer.resize(layer_size, *empty_hash_at(self.ty, layer_i));
             layer_size = max(layer_size >> 1, 1);
         }
-        let start = layers.data[0].len();
         for i in start..new_len {
-            layers.dirt[0].insert(i);
+            let parent_i = i >> 1;
+            assert!(parent_i <= layers.dirty_leaf_parents.len());
+            if parent_i == layers.dirty_leaf_parents.len() {
+                layers.dirty_leaf_parents.push(true);
+            } else if parent_i < layers.dirty_leaf_parents.len() {
+                layers.dirty_leaf_parents.set(parent_i, true);
+            }
         }
         Ok(layers.data[0].len())
     }
 }
 
 impl PartialEq for Merkle {
+    // There are only three members of a Merkle, the type, the layers, and the min_depth.
+    //
+    // It should be obvious that only if the type and layers are equal, will the root hash
+    // be equal. So, it is sufficient to compare the root hash when checking equality.
+    //
+    // However, it is possible that the min_depth may differ between two merkle trees which
+    // have the same type and layers. The root hash will still be equal unless the min_depth
+    // is larger than the depth required to hold the data in the layers.
+    //
+    // For example, a Merkle tree with 5 leaves requires 3 layeers to hold the data. If the
+    // min_depth is 1 on one tree and 2 on another, the root has would still be equal
+    // because the same nodes are hashed together. However, the min_dpeth was 4, then,
+    // there would be 4 layers in that tree, and the root hash would be different.
```

Comment:
One other thing actually, what about two merkle trees of different types? I don't think we should ever be comparing those, but it'd probably be worth adding an `&& self.ty == other.ty`

Reply:
I forgot to mention in this comment, I meant empty merkle trees of different types.

Reply:
Okay. I've added it.

Reply:
No wait. Hold up. This brings up a point that confused me a couple of times as well.

It turns out that empty merkle trees are always of type `MerkleType.Empty` because of these lines:
https://github.com/OffchainLabs/nitro/blob/merkle-perf/arbitrator/prover/src/merkle.rs#L199-L201

BUT ... I'm not sure that makes sense as it violates the principle of least surprise.

For example, if I explicitly tell the API to make a new Merkle tree of type `MerkleType.Value` with an empty vector of leaf hashes, it is going to return a Merkle tree of type `MerkleType.Empty`. But, then, there are a few places in the code that check if the type of the tree is `MerkleType.Empty` and bails. So, maybe this is an important oddity to continue to support. But, it does mean that empty merkle trees will always have the same type and will always be equal to each other no matter what type was passed in during construction.